File: src/go/pkg/infrastructure/oauth/token_source.go
package oauth

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/ripixel/fitglue-server/src/go/pkg/bootstrap"
)

// Token represents the OAuth token structure we care about
type Token struct {
	AccessToken  string
	RefreshToken string
	Expiry       time.Time
}

// TokenSource returns a valid token.
// It is safe for concurrent use by multiple goroutines.
type TokenSource interface {
	Token(context.Context) (*Token, error)
	ForceRefresh(context.Context) (*Token, error)
}

// FirestoreTokenSource reads from Firestore and refreshes if necessary.
type FirestoreTokenSource struct {
	db       *bootstrap.Service
	userID   string
	provider string
	mu       sync.Mutex
}

func NewFirestoreTokenSource(svc *bootstrap.Service, userID, provider string) *FirestoreTokenSource {
	return &FirestoreTokenSource{
		db:       svc,
		userID:   userID,
		provider: provider,
	}
}

// ForceRefresh forcibly refreshes the token regardless of expiry.
func (s *FirestoreTokenSource) ForceRefresh(ctx context.Context) (*Token, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	// 1. Fetch refresh token explicitly from DB again to be safe
	userData, err := s.db.DB.GetUser(ctx, s.userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get user: %w", err)
	}

	if userData.Integrations == nil {
		return nil, fmt.Errorf("user has no integrations linked")
	}

	var refreshToken string
	switch s.provider {
	case "strava":
		if userData.Integrations.Strava == nil || !userData.Integrations.Strava.Enabled {
			return nil, fmt.Errorf("strava not linked/enabled")
		}
		refreshToken = userData.Integrations.Strava.RefreshToken
	case "fitbit":
		if userData.Integrations.Fitbit == nil || !userData.Integrations.Fitbit.Enabled {
			return nil, fmt.Errorf("fitbit not linked/enabled")
		}
		refreshToken = userData.Integrations.Fitbit.RefreshToken
	default:
		return nil, fmt.Errorf("unknown provider %s", s.provider)
	}

	if refreshToken == "" {
		return nil, fmt.Errorf("missing refresh token for %s", s.provider)
	}

	return s.refreshToken(ctx, refreshToken)
}

// Token returns a token, refreshing it if necessary.
func (s *FirestoreTokenSource) Token(ctx context.Context) (*Token, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	// 1. Fetch current token from Firestore
	userData, err := s.db.DB.GetUser(ctx, s.userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get user: %w", err)
	}

	if userData.Integrations == nil {
		return nil, fmt.Errorf("user has no integrations linked")
	}

	var accessToken, refreshToken string
	var expiry time.Time

	switch s.provider {
	case "strava":
		if userData.Integrations.Strava == nil || !userData.Integrations.Strava.Enabled {
			return nil, fmt.Errorf("strava not linked/enabled")
		}
		accessToken = userData.Integrations.Strava.AccessToken
		refreshToken = userData.Integrations.Strava.RefreshToken
		if userData.Integrations.Strava.ExpiresAt != nil {
			expiry = userData.Integrations.Strava.ExpiresAt.AsTime()
		}
	case "fitbit":
		if userData.Integrations.Fitbit == nil || !userData.Integrations.Fitbit.Enabled {
			return nil, fmt.Errorf("fitbit not linked/enabled")
		}
		accessToken = userData.Integrations.Fitbit.AccessToken
		refreshToken = userData.Integrations.Fitbit.RefreshToken
		if userData.Integrations.Fitbit.ExpiresAt != nil {
			expiry = userData.Integrations.Fitbit.ExpiresAt.AsTime()
		}
	default:
		return nil, fmt.Errorf("unknown provider %s", s.provider)
	}

	if accessToken == "" || refreshToken == "" {
		return nil, fmt.Errorf("missing tokens for %s", s.provider)
	}

	// 2. Check Expiry (Proactive Refresh)
	// Refresh if expired or expiring in the next minute
	if !expiry.IsZero() && time.Now().Add(1*time.Minute).After(expiry) {
		return s.refreshToken(ctx, refreshToken)
	}

	return &Token{
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
		Expiry:       expiry,
	}, nil
}

// refreshToken performs the HTTP exchange to get a new token & updates Firestore
func (s *FirestoreTokenSource) refreshToken(ctx context.Context, refreshToken string) (*Token, error) {
	clientID, err := s.getSecret("client-id")
	if err != nil {
		return nil, err
	}
	clientSecret, err := s.getSecret("client-secret")
	if err != nil {
		return nil, err
	}

	var tokenURL string
	switch s.provider {
	case "strava":
		tokenURL = "https://www.strava.com/oauth/token"
	case "fitbit":
		tokenURL = "https://api.fitbit.com/oauth2/token"
	default:
		return nil, fmt.Errorf("unsupported provider for refresh: %s", s.provider)
	}

	data := url.Values{}
	// Strava requires client_id/secret in body. Fitbit uses Basic Auth header (see below).
	if s.provider != "fitbit" {
		data.Set("client_id", clientID)
		data.Set("client_secret", clientSecret)
	}
	data.Set("grant_type", "refresh_token")
	data.Set("refresh_token", refreshToken)

	req, err := http.NewRequestWithContext(ctx, "POST", tokenURL, strings.NewReader(data.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	if s.provider == "fitbit" {
		req.SetBasicAuth(clientID, clientSecret)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("refresh request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("refresh failed with status: %d", resp.StatusCode)
	}

	// Parse Response
	var result struct {
		AccessToken  string `json:"access_token"`
		RefreshToken string `json:"refresh_token"`
		ExpiresIn    int    `json:"expires_in"`
		ExpiresAt    int64  `json:"expires_at"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode refresh response: %w", err)
	}

	newExpiry := time.Now().Add(time.Duration(result.ExpiresIn) * time.Second)
	if result.ExpiresAt != 0 {
		newExpiry = time.Unix(result.ExpiresAt, 0)
	}

	// Update Firestore using UpdateUser logic which expects map (for now)
	// We need snake_case keys for the map update since database.UpdateUser is map-based.
	// We construct a specific nested update manually.
	updateData := map[string]interface{}{
		"integrations." + s.provider + ".access_token":  result.AccessToken,
		"integrations." + s.provider + ".refresh_token": result.RefreshToken,
		"integrations." + s.provider + ".expires_at":    newExpiry,
	}

	if err := s.db.DB.UpdateUser(ctx, s.userID, updateData); err != nil {
		return nil, fmt.Errorf("failed to persist new tokens: %w", err)
	}

	return &Token{
		AccessToken:  result.AccessToken,
		RefreshToken: result.RefreshToken,
		Expiry:       newExpiry,
	}, nil
}

func (s *FirestoreTokenSource) getSecret(keyType string) (string, error) {
	name := fmt.Sprintf("%s-%s", s.provider, keyType)
	return s.db.Secrets.GetSecret(context.Background(), s.db.Config.ProjectID, name)
}

File: src/typescript/shared/src/infrastructure/oauth/token-source.ts
import { UserStore } from '../../storage/firestore';
import { getSecret } from '../secrets/manager';
import { PROJECT_ID } from '../../config';
import { UserIntegrations, FitbitIntegration, StravaIntegration } from '../../types/pb/user';

export interface Token {
  accessToken: string;
  refreshToken: string;
  expiresAt: Date;
}

export interface TokenSource {
  getToken(forceRefresh?: boolean): Promise<Token>;
}

export class FirestoreTokenSource implements TokenSource {
  constructor(
    private userStore: UserStore,
    private userId: string,
    private provider: 'fitbit' | 'strava'
  ) { }

  private getIntegration(integrations: UserIntegrations): FitbitIntegration | StravaIntegration | undefined {
    if (this.provider === 'fitbit') {
      return integrations.fitbit;
    } else if (this.provider === 'strava') {
      return integrations.strava;
    }
    return undefined;
  }

  async getToken(forceRefresh = false): Promise<Token> {
    // 1. Fetch current user from Firestore
    const user = await this.userStore.get(this.userId);
    if (!user) {
      throw new Error(`User ${this.userId} not found`);
    }

    if (!user.integrations) {
      throw new Error(`User ${this.userId} has no integrations`);
    }

    const integration = this.getIntegration(user.integrations);

    if (!integration || !integration.enabled) {
      throw new Error(`${this.provider} integration not enabled for user ${this.userId}`);
    }

    const accessToken = integration.accessToken;
    const refreshToken = integration.refreshToken;
    const expiresAtRaw = integration.expiresAt;

    if (!accessToken || !refreshToken) {
      throw new Error(`Missing tokens for ${this.provider}`);
    }

    // 2. Check Expiry
    // converters.ts guarantees this is a Date or undefined
    const expiresAt = expiresAtRaw || new Date(0);

    // 2. Check Expiry
    const now = new Date();
    const isExpired = expiresAt <= now;

    // Proactive refresh window (1 minute) to match Go implementation
    const isExpiringSoon = expiresAt.getTime() - now.getTime() < 60 * 1000;

    if (forceRefresh || isExpired || isExpiringSoon) {
      return this.refreshTokenFlow(refreshToken);
    }

    return {
      accessToken,
      refreshToken,
      expiresAt
    };
  }

  private async refreshTokenFlow(refreshToken: string): Promise<Token> {
    try {
      const clientId = await getSecret(PROJECT_ID, `${this.provider}-client-id`);
      const clientSecret = await getSecret(PROJECT_ID, `${this.provider}-client-secret`);

      let tokenUrl = '';
      let body: URLSearchParams;
      const headers: HeadersInit = {
        'Content-Type': 'application/x-www-form-urlencoded'
      };

      if (this.provider === 'fitbit') {
        tokenUrl = 'https://api.fitbit.com/oauth2/token';
        const basicAuth = Buffer.from(`${clientId}:${clientSecret}`).toString('base64');
        headers['Authorization'] = `Basic ${basicAuth}`;

        body = new URLSearchParams({
          grant_type: 'refresh_token',
          refresh_token: refreshToken
        });
      } else if (this.provider === 'strava') {
        tokenUrl = 'https://www.strava.com/oauth/token';
        body = new URLSearchParams({
          client_id: clientId,
          client_secret: clientSecret,
          grant_type: 'refresh_token',
          refresh_token: refreshToken
        });
      } else {
        throw new Error(`Unsupported provider for refresh: ${this.provider}`);
      }

      const response = await fetch(tokenUrl, {
        method: 'POST',
        headers: headers,
        body: body
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Refresh failed with status ${response.status}: ${errorText}`);
      }

      const data = await response.json();

      // Normalize response
      const newAccessToken = data.access_token;
      const newRefreshToken = data.refresh_token;
      const expiresIn = data.expires_in; // Seconds

      if (!newAccessToken || !newRefreshToken) {
        throw new Error(`Invalid refresh response from ${this.provider}`);
      }

      const newExpiresAt = new Date(Date.now() + expiresIn * 1000);

      // Fetch current state to merge
      const user = await this.userStore.get(this.userId);
      if (!user || !user.integrations) throw new Error("User lost during refresh");

      // Update Firestore
      const integrationData = user.integrations[this.provider];
      if (!integrationData) {
        throw new Error(`Integration ${this.provider} not found for user ${this.userId} while attempting to update`);
      }
      await this.userStore.setIntegration(this.userId, this.provider, {
        ...integrationData,
        accessToken: newAccessToken,
        refreshToken: newRefreshToken,
        expiresAt: newExpiresAt
      });

      return {
        accessToken: newAccessToken,
        refreshToken: newRefreshToken,
        expiresAt: newExpiresAt
      };

    } catch (error) {
      console.error(`[${this.provider}] Token refresh failed for user ${this.userId}`, error);
      throw error;
    }
  }
}
